### ISSUE TITLE
Support for saving Buffer values

### ISSUE THREAD

**Issue by nonplus** [vitaly-t/pg-promise#105]
Currently when you specify a `Buffer` parameter for a query, it is serialized using `JSON.stringify()` and the JSON representation is saved in the database.  When you query for the value, you get back a `Buffer` containing the JSON representation rather than the original value.

I would expect that when I store a `Buffer` value in a `BYTEA` column, its raw value would be stored.

I think the fix is pretty straight forward, i.e.:

**formatting.js**
``` javascript
function formatValue(value, raw, obj) {
    if (isNull(value)) {
        throwIfRaw(raw);
        return 'null';
    }
    switch (typeof value) {
        ...
        default:
            ...
            if (value instanceof Buffer) {
                return $as.buffer(value, raw);
            }
            ...
    }
}

var $as = {
...
    buffer: function (b, raw) {
        b = resolveFunc(b);
        if (isNull(b)) {
            throwIfRaw(raw);
            return 'null';
        }
        if (b instanceof Buffer) {
            var s = "\\x" + b.toString("hex");
            return raw ? s : TEXT(s);
        }
        throw new TypeError(TEXT(b) + " is not a Buffer object.");
    },
...
}
```
 Would you like me to submit a PR?

**Comment by vitaly-t** [comment]
My understanding of it is that there are multiple ways to insert a `Buffer` into the database, but only one way to read it. A hex string is just one presentation for the `Buffer`.

I need to do a proper research here...

**Comment by vitaly-t** [comment]
@nonplus If I find that there is one way to do it, I will implement it. But I won't be able to accept a PR on the formatting engine, for the following reasons:

* I'm currently working on 3.4.0 with a few changes for the formatting engine, and do not want merging conflicts now.
* `$as.buffer(value, raw);` doesn't adhere to the general logic anyway. First, all formatting methods are named to represent to destination, not the source, and perhaps it wouldn't make sense supporting raw presentation for the Buffer type.

Something to think about.

**Comment by vitaly-t** [comment]
Current findings...

Buffer requires 3 parameters: encoding, start + end. We would have the default to `hex`, 0, buffer length, which I'm not sure if a good thing.

**Comment by nonplus** [comment]
Sorry about the ambiguous case title (I've updated the title) I am actually interested in fixing support for **saving** Buffers to the database. FWIW, **reading** them is handled by `node-postgres` and works just fine.

According to PostgreSQL, the [bytea Hex Format](http://www.postgresql.org/docs/9.5/static/datatype-binary.html#AEN5667):

> The "hex" format encodes binary data as 2 hexadecimal digits per byte, most significant nibble first. The entire string is preceded by the sequence \x (to distinguish it from the escape format). In some contexts, the initial backslash may need to be escaped by doubling it, in the same cases in which backslashes have to be doubled in escape format; details appear below. The hexadecimal digits can be either upper or lower case, and whitespace is permitted between digit pairs (but not within a digit pair nor in the starting \x sequence). The hex format is compatible with a wide range of external applications and protocols, and it tends to be faster to convert than the escape format, so its use is preferred.

This is exactly what my suggested `$as.buffer()` implementation does.
 
Note that all parameters to `Buffer.toString()` are optional ([see documentation](https://nodejs.org/api/buffer.html#buffer_buf_tostring_encoding_start_end)) and the default `start` and `end` values are what we need.

**Comment by vitaly-t** [comment]
I still would have to cater for all the extra formatting overhead, like use of filters `:json` and `:csv`.

Filter `:json` should convert Buffer into a JSON string, and filter `:csv` perhaps should convert it into CSV of byte values - i'm not even sure.

**Comment by nonplus** [comment]
The existing `$as.json` implementation works acceptably with Buffers since it ends up using [Buffer.toJSON()](https://nodejs.org/api/buffer.html#buffer_buf_tojson).  But, I would not expect someone to store a `Buffer` value in an `JSON` (or `JSONB`) column, since it's not a native JSON type and you have to jump through extra hoops when your JSON.parse() it.

Looks like CSV uses backslash characters for encoding binary data, so `$as.csv` would have to account for that (the current implementation converts a Buffer value to its JSON representation).  Though, frankly, trying to use binary data in CSV is not recommended.

For now, I'm not going to worry about these edge cases and fix the Buffer formatting it in my fork.

Thanks for the feedback.

**Comment by vitaly-t** [comment]
I've started making some changes, and changed `as.csv` as this:

```js
    csv: function (values) {
        values = resolveFunc(values);
        if (values instanceof Buffer) {
            return values.toJSON().data.join(',');
        }
        return formatCSV(values);
    },
```

This is just for something consistent, because it is not likely anyone will be using it that way :)

**Comment by nonplus** [comment]
That could work for a single value, but not for an array that contains Buffer(s).  If you look at what Postgres does when it dumps a `BYTEA` column (e.g. `COPY table TO '/tmp/test.csv' WITH CSV HEADER`) you'll see how it encodes the binary values as a string with backslashes.

So you'll probably also want to handle this in **formatCSV**:

``` javascript
function formatCSV(values) {
    if (!values instanceof Array) {
        values = [values];
    }
    return values.map(function (v) {
        if (Buffer.isBuffer(v)) {
            return JSON.stringify(v.toString("binary"));
        }
        return formatValue(v);
    }).join(',');
}
```

**Comment by vitaly-t** [comment]
This would mean that we format `Buffer` type differently from `[Buffer1, Buffer2,...]`, which in turn would make it inconsistent with the rest of the formatting engine.

The question is - what good is `binary` conversion when inserting records? The whole point of the formatting engine is to generate something that can be used directly in queries. And I don't know `binary` can be used in queries.

**Comment by vitaly-t** [comment]
Neither of our approaches would make it right, so after all considerations, the correct approach seems to be this one:

```js
    csv: function (values) {
        values = resolveFunc(values);
        if (values instanceof Buffer) {
            return formatArray(values.toJSON().data);
        }
        return formatCSV(values);
    },
```

**Comment by nonplus** [comment]
```
-- Schema
CREATE TABLE public.person
(
    id SERIAL PRIMARY KEY NOT NULL,
    first VARCHAR(32),
    last VARCHAR(32),
    picture BYTEA
);

// Insert a person with an image
ctx.none("INSERT INTO person SET (first, last, picture) VALUES (${first}, ${last}, ${picture})", {
   first: "Georger",
   last: "Washington",
   picture: new Buffer(...)
});
```

**Comment by nonplus** [comment]
BTW, what _is_ the use case for the CSV formatting?  I'm familiar with bulk imports/exports (`COPY FROM/TO`), but I've only used those with CSV files.

**Comment by vitaly-t** [comment]
Franky, none, perhaps. But if one calls and we do not support - we need to throw an error of some sort. I think that somehow logical support is better than adding new types of errors in the midst :)

**Comment by vitaly-t** [comment]
@nonplus initial implementation + 100% tests have been added: https://github.com/vitaly-t/pg-promise/commit/b76c85d2aea4b381898ecaee4484999ccc56f7e8

**Comment by vitaly-t** [comment]
There was a build error, because as it turns out, Buffer was converted into JSON differently under Node.js. I have fixed that and check it in again ;)

**Comment by nonplus** [comment]
Looks good.  I think the CSV case is inconsistent with the other types where `someValue` and `[someValue]` result in the same CSV representation.  In any case, you shouldhave a CSV test case where Buffer is an array element.
```
expect(pgp.as.csv([23, new Buffer([1, 2, 3]), "Hello")).toBe(???);
```

If it were me, I would make `as.csv` stricter and have it throw errors if caller doesn't pass in an array, or if any of the array elements aren't primitive values (null, Number, String, Boolean, Date, Buffer) since CSV doesn't support structured values (arrays, objects).  Alternatively, you could `JSON.stringify()` the structured values but store them as **properly escaped strings**.

Right now, (from your unit test) you convert `[1, [2, 3], 4]` to `"1,array[2,3],4"` which is actually the CSV representation of `[1, "array[2", "3]", 4]`. Was that your intention? I would either expect an error or `1,'[2,3]',4`

BTW, CSV is encoding binary characters as as 3-digit octal sequence.  The following function should convert a Buffer to its CSV representation:
```
function bufferToCsv(buf) {
	return "'" + buf.toJSON().data.map((b) => {
		var c = String.fromCharCode(b);
		if (b >= 32 && b < 127 && c != "'" && c != "\"" && c != ",") {
			return c;
		}
		return "\\" + ("00" + b.toString(8)).substr(-3);
	}).join("") + "'";
}

// Buffer test
expect(pgp.as.csv([
    23,
    new Buffer([ 0, 1, 72, 101, 108, 108, 111, 255 ]),
    "World"
])).toBe("23,'\\000\\001Hello\\377',World");
```

**Comment by vitaly-t** [comment]
>I think the CSV case is inconsistent with the other types where someValue and [someValue] result in the same CSV representation.

I believe it is consistent now. Any example?

>In any case, you shouldhave a CSV test case where Buffer is an array element.

I did include this into the tests, see https://github.com/vitaly-t/pg-promise/blob/master/test/formatSpec.js#L29

>If it were me, I would make as.csv stricter and have it throw errors if caller doesn't pass in an array

That would make it a huge breaking change, because all function calls rely on `as.csv` that is often used with a single parameter.

>or if any of the array elements aren't primitive values (null, Number, String, Boolean, Date, Buffer) since CSV doesn't support structured values (arrays, objects)

Can't do it, method `as.csv` is used to format values for function calls.

>Right now, (from your unit test) you convert [1, [2, 3], 4] to "1,array[2,3],4" which is actually the CSV representation of [1, "array[2", "3]", 4]. Was that your intention? I would either expect an error or 1,'[2,3]',4

As explained earlier, if of the main purposes of the method is to format parameters for function calls.

>BTW, CSV is encoding binary characters as as 3-digit octal sequence. The following function should convert a Buffer to its CSV representation:

The problem, as we discussed it earlier, the purpose of such formatting is unclear.

**Comment by vitaly-t** [comment]
If I were to throw in that exact test, it would work like this:
```js
expect(pgp.as.csv([23, new Buffer([1, 2, 3]), "Hello"])).toBe("23,'\\x010203','Hello'");
```
I do it in parts instead, but I tested it, and that's how it works ;)

**Comment by vitaly-t** [comment]
@nonplus here I have checked in the additional tests: https://github.com/vitaly-t/pg-promise/commit/30fe63ce697b99adb13297e80472cd0d32c98a6d

**Comment by vitaly-t** [comment]
Version [3.4.0](https://github.com/vitaly-t/pg-promise/releases/tag/v.3.4.0) has been released, with all the new features as planned.

I also updated the example [here](https://github.com/vitaly-t/pg-promise/wiki/Learn-by-Example#insert-binary) for using type `Buffer`.

If you find any issues with support for `Buffer`, feel free to open a new issue, and reference this one. But for now the feature has been finished, and it seems to work fine, with a comprehensive set of tests to support that.
